File: new-api-tester/src/lib/dates.js
/**
 * A calendar day in a box, an instant on the wire.
 *
 * The API deals in instants — `2027-03-31T23:59:59Z` — and people deal in days. A date input
 * gives a day and a calendar to pick it from; these four turn one into the other.
 *
 * ONE COPY, BECAUSE THE END-OF-DAY RULE IS EASY TO GET WRONG. It was written twice before this
 * module existed, and two copies of a rule about billing dates is one copy that eventually says
 * midnight where the other says a second to midnight.
 *
 * EVERYTHING HERE IS UTC, matching the `Z` the API sends back, so a day picked in the box is the
 * same day when it is read again. The billing period the backend derives from a plan's cycle is
 * calculated in the school's own zone; that is a different question from what a picker sends, and
 * anywhere the distinction matters the screen shows the stored instant next to the box.
 */

/** An instant from the API as the yyyy-MM-dd a date input wants. "" for anything unreadable. */
export function toDateInput(instant) {
  if (!instant) return ''
  const at = new Date(instant)
  return Number.isNaN(at.getTime()) ? '' : at.toISOString().slice(0, 10)
}

/** Today, in the same shape, so a form's default matches the API's. */
export function todayInput() {
  return new Date().toISOString().slice(0, 10)
}

/** The start of a chosen day, or null for blank. "From the 1st" includes the whole 1st. */
export function startOfDay(value) {
  return value ? `${value}T00:00:00Z` : null
}

/**
 * The end of a chosen day, or null for blank.
 *
 * END OF DAY, NOT MIDNIGHT. "Stops being sold on 31 March" means the 31st is the last day it can
 * be sold, and midnight on the 31st would cut it off before that day started — it would close a
 * day earlier than whoever typed it expected.
 */
export function endOfDay(value) {
  return value ? `${value}T23:59:59Z` : null
}

/**
 * The instant a given day begins in a given IANA zone.
 *
 * NEEDED BECAUSE #13 MADE currentPeriodStart COMPULSORY. Before that the form could omit it and
 * let the API default to "the start of today in the school's zone"; now the form has to name that
 * instant, and `${day}T00:00:00Z` is the wrong one for any school not on UTC. For Asia/Kolkata
 * the day begins at 18:30Z the evening before; for America/New_York, at 04:00Z the same morning —
 * and that second case is what a naive UTC midnight gets refused for, because 00:00Z is still the
 * previous day there and the API reads it as a past start.
 *
 * Falls back to UTC midnight when the zone is unknown or unusable, which is the best guess
 * available and no worse than what the form sent before.
 */
export function startOfDayInZone(day, zone) {
  if (!day) return null
  const utcMidnight = Date.parse(`${day}T00:00:00Z`)
  if (Number.isNaN(utcMidnight)) return null
  if (!zone) return `${day}T00:00:00Z`

  try {
    // What UTC midnight of that day reads as in the zone, used to recover the zone's offset then.
    // Two passes, because the offset can itself differ either side of midnight on a DST boundary.
    let guess = utcMidnight
    for (let pass = 0; pass < 2; pass += 1) {
      const parts = new Intl.DateTimeFormat('en-GB', {
        timeZone: zone, hour12: false,
        year: 'numeric', month: '2-digit', day: '2-digit',
        hour: '2-digit', minute: '2-digit', second: '2-digit',
      }).formatToParts(new Date(guess))
      const at = Object.fromEntries(parts.filter((p) => p.type !== 'literal')
        .map((p) => [p.type, Number(p.value)]))
      const asIfUtc = Date.UTC(at.year, at.month - 1, at.day,
        at.hour % 24, at.minute, at.second)
      guess = utcMidnight - (asIfUtc - guess)
    }
    return new Date(guess).toISOString().replace('.000', '')
  } catch {
    // An unknown zone throws in Intl. UTC midnight is the honest fallback.
    return `${day}T00:00:00Z`
  }
}

/** Today's date as it stands in a given zone, which is not always today's UTC date. */
export function todayInZone(zone) {
  if (!zone) return todayInput()
  try {
    return new Intl.DateTimeFormat('en-CA', {
      timeZone: zone, year: 'numeric', month: '2-digit', day: '2-digit',
    }).format(new Date())
  } catch {
    return todayInput()
  }
}

/**
 * An instant as a person reads it: "Friday 8 October 2027 11:59 PM".
 *
 * THE PROJECT RULE FOR ANYTHING ON SCREEN. A raw `2027-10-08T23:59:59Z` in a table row is
 * correct and nobody reads it, and the backend already spells out every date inside its
 * messages — a screen that prints the raw field beside one of those messages undoes that.
 *
 * THE TIME IS ALWAYS SHOWN, even when it looks redundant, for the same reason the backend shows
 * it: these dates get compared against each other and both ends can fall on the same day.
 *
 * THE LOCALE IS PINNED, not taken from the browser. `undefined` makes the format a property of
 * whoever opened the page, and a screenshot in a bug report has to mean the same thing as what
 * the next person sees.
 *
 * UTC, matching the `Z` the API sends. The school's own zone would name a different day for the
 * same instant — see the note at the top of this module — and that choice is not settled; where
 * it matters, the screen shows the stored instant next to this via `readableInstant`.
 */
export function readableDateTime(instant) {
  if (!instant) return '—'
  const at = new Date(instant)
  if (Number.isNaN(at.getTime())) return String(instant)

  const parts = new Intl.DateTimeFormat('en-GB', {
    timeZone: 'UTC', weekday: 'long', day: 'numeric', month: 'long', year: 'numeric',
    hour: 'numeric', minute: '2-digit', hour12: true,
  }).formatToParts(at)
  const at_ = Object.fromEntries(parts.filter((p) => p.type !== 'literal')
    .map((p) => [p.type, p.value]))

  // Assembled rather than taken from the formatter's own joining, so the shape is ours: no comma
  // after the weekday, and AM/PM upper case with one space before it.
  return `${at_.weekday} ${at_.day} ${at_.month} ${at_.year} `
    + `${at_.hour}:${at_.minute} ${(at_.dayPeriod || '').toUpperCase()}`
}

/**
 * The exact stored instant, unchanged.
 *
 * KEPT ALONGSIDE `readableDateTime` ON PURPOSE, and it is not a violation of the readable-date
 * rule. This is an API testing tool: next to a date picker that can only show a day, the thing
 * somebody needs to see is the precise value on the wire, including the seconds and the `Z`.
 * Use this for "stored as ..." hints; use `readableDateTime` for anything presented as
 * information.
 */
export function readableInstant(instant) {
  if (!instant) return null
  const at = new Date(instant)
  return Number.isNaN(at.getTime()) ? String(instant) : at.toISOString().replace('.000', '')
}

File: backend/src/main/java/com/orbitastra/backend/common/time/Dates.java
package com.orbitastra.backend.common.time;

import java.time.DateTimeException;
import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneId;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.Locale;

/**
 * Dates as a person reads them, for anything a person is going to read.
 *
 * <pre>
 * Dates.readable(Instant.parse("2027-10-08T16:31:00Z"), "Asia/Kolkata")
 *     -> "Friday 8 October 2027 10:01PM"
 * </pre>
 *
 * <p><b>Why this exists.</b> An {@code Instant} concatenated into a string prints its ISO-8601
 * form — {@code 2027-10-08T23:59:59Z} — and that was going into API messages a human being reads:
 * <i>"SUB/2026/09/000002 runs to 2027-10-08T23:59:59Z, which has not passed yet"</i>. Correct, and
 * nobody reads it. Every message that names a date now comes through here, so the format is one
 * decision in one place rather than thirty string concatenations that drift apart.
 *
 * <p><b>THE ZONE IS NOT COSMETIC — it decides the calendar day.</b> A billing period starts at
 * midnight in the school's own timezone, which is stored as an instant 5½ hours earlier for an
 * Indian school:
 *
 * <pre>
 * 2026-09-07T18:30:00Z  in UTC          -> "Monday 7 September 2026 6:30PM"
 * 2026-09-07T18:30:00Z  in Asia/Kolkata -> "Tuesday 8 September 2026 12:00AM"   &lt;- what the school means
 * </pre>
 *
 * <p>So <b>pass the school's zone whenever there is a school in hand</b>. Rendering that instant in
 * UTC would tell a school its period began on the 7th when its own calendar says the 8th, and
 * being off by a day is worse than being unreadable. The no-zone overload exists for the platform
 * catalogue, where a plan's selling window belongs to no school and UTC is the honest answer.
 *
 * <p><b>THE TIME IS ALWAYS SHOWN</b>, even when it looks redundant. Two of these messages compare
 * one date against another — "must be after", "is before the start of today" — and both ends can
 * fall on the same day. Dropping the time there produces <i>"(8 September 2026) must be after
 * (8 September 2026)"</i>, which reads as a contradiction and tells the caller nothing about what
 * to change.
 *
 * <p><b>{@code Locale.ENGLISH} is pinned deliberately, not left to the JVM.</b> The default locale
 * changes the output: {@code en_IN} renders "10:01pm" in lower case, and a JVM started elsewhere
 * would render the month name in another language. An API message is part of the contract, so it
 * cannot depend on how the process was launched.
 */
public final class Dates {

    /**
     * "Friday 8 October 2027 10:01PM". The weekday is included because these dates decide whether
     * a school is working, and a bare number does not say that.
     */
    private static final DateTimeFormatter WITH_TIME =
            DateTimeFormatter.ofPattern("EEEE d MMMM yyyy h:mma", Locale.ENGLISH);

    /** "Friday 8 October 2027". A LocalDate has no time to show, so none is invented. */
    private static final DateTimeFormatter DATE_ONLY =
            DateTimeFormatter.ofPattern("EEEE d MMMM yyyy", Locale.ENGLISH);

    /**
     * What a null date reads as.
     *
     * <p>Sooner said than printed: several of these values are genuinely nullable — a subscription
     * mid-migration with no period end, a plan with no closing date — and concatenating one gave
     * the literal string "null" in a message somebody had to act on.
     */
    private static final String NOT_SET = "(not set)";

    private Dates() {
    }

    /**
     * An instant as the school's own calendar sees it.
     *
     * <p>{@code zone} is the school's {@code defaultTimeZone}. An unset or unrecognisable one
     * falls back to UTC rather than throwing: a message explaining a refusal must not itself fail,
     * and a wrong-by-hours date still tells the caller more than a 500 does.
     */
    public static String readable(Instant instant, String zone) {
        if (instant == null) {
            return NOT_SET;
        }

        return WITH_TIME.format(instant.atZone(zoneOrUtc(zone)));
    }

    /**
     * An instant in UTC, for the platform surface where no school owns the date.
     *
     * <p>Use the two-argument form wherever a school is in scope — see the note on this class about
     * the calendar day. This one is for plan selling windows and anything else the platform sets
     * for itself.
     */
    public static String readable(Instant instant) {
        return readable(instant, null);
    }

    /**
     * A calendar date, which is already a day and needs no zone.
     *
     * <p>Academic year and holiday dates are stored as {@code LocalDate} precisely because they
     * are days rather than moments. Converting one to an instant to print a time would invent
     * information.
     */
    public static String readable(LocalDate date) {
        if (date == null) {
            return NOT_SET;
        }

        return DATE_ONLY.format(date);
    }

    /**
     * Midnight today, as the school's own calendar sees it.
     *
     * <p>"Today" is not a fixed instant — it depends whose day you mean. Midnight in
     * Asia/Kolkata is 18:30Z the evening before, so a check like "is this date in the past"
     * answered in UTC would call a school's own today yesterday.
     *
     * <p>Here rather than in a service, because it is the same question {@link #readable} already
     * answers: which day is this, for this school. It reuses the same zone fallback, so a
     * rendered date and a date comparison can never disagree about the zone.
     */
    public static Instant startOfTodayIn(String zone) {
        ZoneId resolved = zoneOrUtc(zone);

        return LocalDate.now(resolved).atStartOfDay(resolved).toInstant();
    }

    /**
     * Today's calendar date, as the school's own calendar sees it.
     *
     * <p>Beside {@link #startOfTodayIn} because it answers the same question for the other date
     * type. An academic year's boundaries are {@code LocalDate} — they are days, not moments —
     * so "is today inside this year" cannot be asked with an {@code Instant} without converting
     * one of them and inventing information in the process.
     *
     * <p><b>The zone is not cosmetic here either.</b> At 23:00 in Asia/Kolkata it is still the
     * previous day in UTC, so a year that ends today would already read as finished. Reuses the
     * same zone fallback as everything else in this class, so a rendered date and a date
     * comparison can never disagree about the zone.
     */
    public static LocalDate todayIn(String zone) {
        return LocalDate.now(zoneOrUtc(zone));
    }

    /**
     * The school's zone, or UTC when it is unset or unusable.
     *
     * <p>Inline in each method rather than shared with the callers that also need a ZoneId: this
     * one is about rendering, and a formatter that quietly failed on a bad zone string would take
     * a refusal message down with it.
     */
    private static ZoneId zoneOrUtc(String zone) {
        if (zone == null || zone.isBlank()) {
            return ZoneOffset.UTC;
        }

        try {
            return ZoneId.of(zone.trim());
        } catch (DateTimeException e) {
            return ZoneOffset.UTC;
        }
    }
}
